bin/node/src/pruning_config.rs

```diff
@@ -0,0 +1,151 @@
+use log::warn;
+use primitives::DEFAULT_SESSION_PERIOD;
+use sc_cli::{Database, DatabasePruningMode, PruningParams};
+use static_assertions::const_assert;
+
+use crate::Cli;
+
+/// Anything greater than 1800, which is two normal length sessions, should be enough.
+/// We need to be able to read back so many states to retrieve the list of authorities for a session.
+const MINIMAL_STATE_PRUNING: u32 = 2048;
+const_assert!(MINIMAL_STATE_PRUNING >= 2 * DEFAULT_SESSION_PERIOD);
+
+const DEFAULT_STATE_PRUNING: DatabasePruningMode = DatabasePruningMode::Archive;
+
+const DEFAULT_BLOCKS_PRUNING: DatabasePruningMode = DatabasePruningMode::ArchiveCanonical;
+
+const DEFAULT_DATABASE_FOR_PRUNING: sc_cli::Database = Database::ParityDb;
+
+pub struct PruningConfigValidator {
+    pruning_enabled: bool,
+    overwritten_pruning: bool,
+    invalid_state_pruning_setting: Result<(), u32>,
+    invalid_blocks_pruning_setting: Result<(), u32>,
+    invalid_database_backend: Result<(), ()>,
+}
+
+impl PruningConfigValidator {
+    fn pruning_changed(params: &PruningParams) -> bool {
+        let state_pruning_changed =
+            params.state_pruning.unwrap_or(DEFAULT_STATE_PRUNING) != DEFAULT_STATE_PRUNING;
+        let blocks_pruning_changed = params.blocks_pruning != DEFAULT_BLOCKS_PRUNING;
+
+        state_pruning_changed || blocks_pruning_changed
+    }
+
+    pub fn validate_and_fix_pruning_options(cli: &mut Cli) -> Self {
+        let overwritten_pruning = Self::pruning_changed(&cli.run.import_params.pruning_params);
+        let pruning_enabled = cli.aleph.enable_pruning();
+
+        let mut result = PruningConfigValidator {
+            pruning_enabled,
+            overwritten_pruning,
+            invalid_state_pruning_setting: Ok(()),
+            invalid_blocks_pruning_setting: Ok(()),
+            invalid_database_backend: Ok(()),
+        };
+
+        if !pruning_enabled {
+            // We need to override state pruning to our default (archive), as substrate has 256 by default.
+            // 256 does not work with our code.
+            cli.run.import_params.pruning_params.state_pruning = Some(DEFAULT_STATE_PRUNING);
+            cli.run.import_params.pruning_params.blocks_pruning = DEFAULT_BLOCKS_PRUNING;
+            return result;
+        }
+
+        match cli
```

Comment:
Nit: it would be slightly nicer if these `match`es were methods called here by name.

Reply:
moved